akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2015-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.annotation.InternalApi
+import akka.stream.scaladsl.{ Flow, Keep, Source }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.stream._
+import akka.util.OptionVal
+
+import scala.collection.immutable
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+@InternalApi private[akka] final class FlatMapPrefix[In, Out, M](n: Int, f: immutable.Seq[In] => Flow[In, Out, M])
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+
+  require(n >= 0, s"FlatMapPrefix: n ($n) must be non-negative.")
+
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      var accumulated = List.empty[In]
+
+      private var subSource = OptionVal.none[SubSourceOutlet[In]]
+      private var subSink = OptionVal.none[SubSinkInlet[Out]]
+
+      setHandlers(in, out, this)
+
+      override def postStop(): Unit = {
+        matPromise.tryFailure(new AbruptStageTerminationException(this))
+        if (subSource.isDefined && !subSource.get.isClosed)
+          subSource.get.complete()
```

Comment:
Should that source be failed there?

Reply:
hmm... this is abrubt-termination land, we might be able to rely on the interpreter calling post-stop on the materialized flow as well. this way we'd get the materialized flow's behaviour: either failing or completing, I'm just not sure if we'd get the future matval properly completed...
will have to look a bit deeper into this

Reply:
thinking about this a bit further:
substop can be calles in one of two cases: 
1. after stage completion, hence all ports (in and out) were closed.
2. abrupt termination.

case 1 implies that onUpstreamCompletion/failure and onDownstreamCompletion(failure) were invoked, hence we had our chance of notifying the materialized flow of this events, hence in this case we should **not** close the ports during postStop.
case 2 implies that the materialized flow will soon be stopped as well (if not already stopped), hence again - no need to close its ports.

@johanandren , @jrudolph , does this make any sense to you guys?

Reply:
The subflow will always be same materializer of this graphstage so I think it'd be safe to rely on that abruptly terminating the nested flow as well.